src/cpp/shared/combat/army.cpp

```diff
@@ -0,0 +1,1626 @@
+#include "artifacts.h"
+#include "combat/army.h"
+#include "combat/combat.h"
+#include "resource/resourceManager.h"
+#include "scripting/callback.h"
+#include "sound/sound.h"
+#include "spells.h"
+
+extern char *cMonFilename[]; // it's inside creature.cpp
+extern char *cArmyFrameFileNames[]; // it's inside creature.cpp
+extern char *cArmyProjectileFileNames[]; // it's inside creature.cpp
+extern icon *gCurLoadedSpellIcon;
+extern int gCurLoadedSpellEffect;
+extern int gCurSpellEffectFrame;
+extern int gbGenieHalf;
+extern int gbRemoteOn;
+extern int giWalkingFrom;
+extern int giWalkingFrom2;
+extern int giWalkingTo;
+extern int giWalkingTo2;
+extern int giWalkingYMod;
+extern int gbComputeExtent;
+extern int gbSaveBiggestExtent;
+extern int gbReturnAfterComputeExtent;
+extern int gbCurrArmyDrawn;
+extern int gbLimitToExtent;
+extern unsigned __int8 moatCell[];
+extern float gfBattleStat[];
+extern float gfSSArcheryMod[];
+
+bool gCloseMove; // ironfist var to differentiate between close/from a distance attack
+
+char *gCombatFxNames[33] =
+{
+  "",
+  "magic01.icn",
+  "magic02.icn",
+  "magic03.icn",
+  "magic04.icn",
+  "magic05.icn",
+  "magic06.icn",
+  "magic07.icn",
+  "magic08.icn",
+  "rainbluk.icn",
+  "cloudluk.icn",
+  "moraleg.icn",
+  "moraleb.icn",
+  "reddeath.icn",
+  "redfire.icn",
+  "sparks.icn",
+  "electric.icn",
+  "physical.icn",
+  "bluefire.icn",
+  "icecloud.icn",
+  "lichclod.icn",
+  "bless.icn",
+  "berzerk.icn",
+  "shield.icn",
+  "haste.icn",
+  "paralyze.icn",
+  "hypnotiz.icn",
+  "dragslay.icn",
+  "blind.icn",
+  "curse.icn",
+  "stonskin.icn",
+  "stelskin.icn",
+  "plasmblast.icn"
+};
+
+unsigned __int8 giNumPowFrames[33] =
+{
+  10u,
+  10u,
+  10u,
+  10u,
+  10u,
+  10u,
+  10u,
+  10u,
+  10u,
+  8u,
+  8u,
+  10u,
+  10u,
+  10u,
+  10u,
+  15u,
+  10u,
+  10u,
+  10u,
+  10u,
+  10u,
+  16u,
+  16u,
+  14u,
+  19u,
+  22u,
+  10u,
+  17u,
+  10u,
+  12u,
+  11u,
+  16u,
+  7u
+};
+
+void OccupyHexes(army *a) {
+  if (!(a->creature.creature_flags & TWO_HEXER))
+    return;
+
+  if (a->facingRight == 1)
+    a->occupiedHex--;
+  else
+    a->occupiedHex++;
+}
+
+int __fastcall OppositeDirection(signed int hex) {
+  int result;
+  if (hex == 6) {
+    result = 7;
+  } else if (hex > 6) {
+    result = 6;
+  } else {
+    result = (hex + 3) % 6;
+  }
+  return result;
+}
+
+void DoAttackBattleMessage(army *attacker, army *target, int creaturesKilled, int damageDone) {
```

Comment:
Would the code be shorter? Afaik we can simulate sprintf format function only with std string streams. And those aren't very beatiful